Add ModelFeatureFlag and improve plugin support

vojtechszocs opened the pull request:
## New extensions

This PR adds an extension type that allows Console plugins to contribute new model-based feature flags.

For example:

```ts
const plugin: Plugin<ModelFeatureFlag> = [
  {
    type: 'FeatureFlag/Model',
    properties: {
      model: VirtualMachineModel,
      flag: 'KUBEVIRT',
    },
  },
];
```

In general, every Console plugin should contribute one (or possibly more) feature flags related to that plugin. Consequently, any extension of Console UI should be gated by the corresponding feature flag at runtime.

So if e.g. KubeVirt is not detected on the cluster, any KubeVirt related extensions will not be effective.

## Code improvements

- `@console/active-plugins` codegen simplified and unit-tested

![test-codegen](https://user-images.githubusercontent.com/648971/57157096-01b12800-6de0-11e9-8cee-bc027ac8c182.png)

- using TypeScript [namespaces](https://www.typescriptlang.org/docs/handbook/namespaces.html) to better organize extension's `properties` types, for example:

```ts
namespace ExtensionProperties {
  export interface ModelFeatureFlag {
    model: K8sKind;
    flag: string;
  }
}

export interface ModelFeatureFlag extends Extension<ExtensionProperties.ModelFeatureFlag> {
  type: 'FeatureFlag/Model';
}
```

- `eslint-plugin-import` bumped to v2.17.2 in order to [support TypeScript namespaces](https://github.com/benmosher/eslint-plugin-import/pull/1320)
- suppress `<N> plugins active` log when `NODE_ENV` is `test` (avoid test output clutter)
- print the list of active plugins during webpack build

![log-active-plugins](https://user-images.githubusercontent.com/648971/57157311-a6336a00-6de0-11e9-898e-33751efaba28.png)

---

/cc @spadgett @christianvogt @jelkosz

vojtechszocs commented:
Follow-up to #1499

vojtechszocs commented:
/cc @alecmerdler

alecmerdler commented on frontend/packages/console-plugin-sdk/src/codegen/index.test.ts:
```diff
@@ -0,0 +1,104 @@
+import {
```
Why aren't we following the `module.spec.ts` filename pattern here?

vojtechszocs replied:
You're right :+1: we should follow the existing pattern.

I'll do the change and update Jest `testRegex` value accordingly.

@alecmerdler @spadgett What's your opinion on unit test co-location? Right now, the test sits next to the corresponding module. An alternative is to have e.g.

```
packages/console-plugin-sdk/__tests__/codegen/index.spec.ts
```

which adds structural duplicity in favor of better test isolation.

Once we move code out of the `public` directory into appropriate packages, which test location convention should we follow?

alecmerdler replied:
I prefer having tests next to their source file, but that's not what we had when I started working on `bridge`, so we've stuck with the `__tests__` directory. We should stick with the duplicate structure; following the `module.spec.ts` filename format makes it easy to find the test files using search anyway.

christianvogt replied:
Why not just follow the typical pattern found and described by most react apps and tutorials wherein the `__tests__` directory is located adjacent to the file being tested. So in this case create test file at `src/codegen/__tests__/index.spec.ts`

vojtechszocs replied:
> I prefer having tests next to their source file

I have the same preference, the main reason being that it's immediately clear if the given module has a corresponding test, and the secondary reason being that it encourages developers to treat tests in the same way as actual code.

I'm OK with @christianvogt's proposal, co-locating tests under `__tests__` adjacent to the module being tested. Should I proceed with this change? @spadgett @alecmerdler 

The underscores in `__tests__` is something I don't like, though. I guess this comes from Python, which uses `__` prefix to denote things that are outside of the regular code. (From my point of view, unit tests should be an integral part of the regular code, not something outside of it.)

alecmerdler replied:
If we are going to co-locate the tests, my vote would to have a flat structure and not use `__tests__`. So: 
```
components/
|--- index.ts
|--- index.spec.ts
|--- widget.tsx
|--- widget.spec.tsx
```
Whatever we choose, we should do it across the entire codebase to ensure consistency, which means breaking out the existing `__tests__` directory to their respective source directories.

vojtechszocs replied:
> Whatever we choose, we should do it across the entire codebase to ensure consistency

I'll take care of that as part of the "moving stuff out of `public` directory" effort.

christianvogt replied:
I agree that we should do it across the whole repo. But do so in steps.
One benefit to the `__tests__` dir is that related test files can also be placed in there as to not confused with application code. The double underscores help move the dir to the top of the file sort order. I can appreciate having tests as sibling files though.

When using jest we also have a `__mocks__` folder for mocks and a `__snapshots__` folder for snapshots. Snapshots could probably be co-located once again in the same dir. TBH I never looked at changing the directory for mocks though.

Something worth bringing up as we look to improve and enhance the coding guidelines as well as look to add more linting.

vojtechszocs replied:
> One benefit to the `__tests__` dir is that related test files can also be placed in there as to not confused with application code. The double underscores help move the dir to the top of the file sort order. I can appreciate having tests as sibling files though.

These are good points. @spadgett do you have any thoughts on this?

For this PR, I'll use the adjacent `__tests__` directory convention:

```
path/to/module.ts(x)
path/to/__tests__/module.spec.ts(x)
```

> When using jest we also have a `__mocks__` folder for mocks and a `__snapshots__` folder for snapshots.

Actually, some files under the `__mocks__` directory are [setup files](https://jestjs.io/docs/en/configuration.html#setupfiles-array):

```
"setupFiles": [
  "./__mocks__/requestAnimationFrame.js",
  "./__mocks__/localStorage.ts",
  "./__mocks__/matchMedia.js",
  "./before-tests.js"
],
```

Some other files in that directory are [automatic module mocks](https://jestjs.io/docs/en/configuration.html#modulenamemapper-object-string-string):

```
"moduleNameMapper": {
  "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.js",
  "\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js"
},
```

The remaining files in that directory are what I call helper mocks. These are used in tests to provide mocked versions of specific objects, for example:

```
// in __tests__/components/catalog.spec.tsx
import { catalogListPageProps, catalogItems, catalogCategories } from '../../__mocks__/catalogItemsMocks';
```

As far as I can see, Console doesn't use Jest [manual mocks](https://jestjs.io/docs/en/manual-mocks) which are used to stub out entire modules, both application and vendor.

> Snapshots could probably be co-located once again in the same dir. TBH I never looked at changing the directory for mocks though.

Jest originally introduced the `__snapshots__` directory for the purpose of completely unambiguous test vs. snapshot file resolution.

As of Jest 24, it's [possible to customize the test snapshot location](https://github.com/facebook/jest/issues/1650#issuecomment-435545153) by providing a custom [snapshot resolver](https://jestjs.io/docs/en/configuration.html#snapshotresolver-string).

I don't see any value in having many `__snapshots__` directories dispersed across the codebase, so I'd vote for co-locating test snapshot right next to the test itself. If we go with the adjacent `__tests__` convention, this means:

```
path/to/module.ts(x)
path/to/__tests__/module.spec.ts(x)
path/to/__tests__/module.spec.snap
```

vojtechszocs replied:
Also, by putting test snapshots next to the corresponding tests,

> related test files can also be placed in there as to not confused with application code

makes even more sense.

christianvogt commented on frontend/package.json:
```diff
@@ -36,9 +37,9 @@
+      "<rootDir>/node_modules/(?!lodash-es|@console)"
     ],
-    "testRegex": "/__tests__/.*\\.(ts|tsx|js|jsx)$",
+    "testRegex": "(/__tests__/.*|\\.test)\\.(ts|tsx|js|jsx)$",
```
Stick to using `spec` as that's the adopted file name pattern

vojtechszocs replied:
Agreed, will do.

vojtechszocs commented:
Rebased & updated according to review comments.

vojtechszocs commented on frontend/packages/console-plugin-sdk/src/typings/features.ts:
```diff
@@ -0,0 +1,24 @@
+}
+
+// TODO(vojtech): add ActionBasedFeatureFlag
+export type FeatureFlag = ModelBasedFeatureFlag;
```
Sounds good, I'll do the change.

spadgett replied:
We'll probably want something like an `RBACFeatureFlag` eventually. I've been investigating ways of handling RBAC in console generally to conditionally disable actions and nav items. I think we can add it when we need it.

vojtechszocs commented:
PR updated :man_technologist: all review comments should be addressed.

spadgett commented:
/retest

spadgett commented:
/retest

vojtechszocs commented:
> @vojtechszocs: PR needs rebase.

I'm on it.

vojtechszocs commented:
@spadgett @alecmerdler Currently, there's `const CRDs` in both `public/actions/features.ts` and `public/reducers/features.ts`, however the first one seems unused.

So I've updated the second one with

```
[referenceForModel(MachineAutoscalerModel)]: FLAGS.MACHINE_AUTOSCALER,
```

and removed the first one :smiley:

vojtechszocs commented:
Fixed the failing `__tests__/reducers/features.spec.tsx`, no other changes.

spadgett commented:
/lgtm
